sips/pending/_posts/2017-01-13-binary-compatibility.md

```diff
@@ -0,0 +1,168 @@
+ 
+The limitations enforced by `@binaryCompatible` are designed to be an overapproximation: 
+instead of permitting a list of features known to be compatible, `@binaryCompatible` enforces a stronger 
+check which is sufficient to promise binary compatibility. 
+
+## Overview ##
+In order for a class or a trait to succeed compilation with the `@binaryCompatible` annotation it has to be:
+  - defined on the top level;
+  - use a subset of Scala that during compilation does not require changes to public API of the class, including
+     - synthesizing new members, either concrete or abstract;
+     - changing binary signatures of existing members, either concrete or abstract;
+
+`@binaryCompatible` does not change the compilation scheme of a class:
+ compiling a class previously annotated with the `@binaryCompatible`, will produce the same bytecode with or without `@binaryCompatible` annotation. 
+
+Below are several examples of classes and traits that succeed compilation with `@binaryCompatible`
+```scala
+{% highlight scala %}
+@binaryCompatible
+trait AbstractFile {
+  def name(): String
+
+  def path(): String
+
+  def jfile(): Optional[File]
+}
+
+@binaryCompatible
+trait SourceFile extends AbstractFile {
+  def content(): Array[Char]
+}
+
+@binaryCompatible
+trait Diagnostic {
+  def message(): String
+
+  def level(): Int
+
+  def position(): Optional[SourcePosition]
+}
+
+@binaryCompatible
+object Diagnostic {
+  @static final val ERROR: Int = 2
+  @static final val WARNING: Int = 1
+  @static final val INFO: Int = 0
+}
+
+@binaryCompatible
+class FeaturesInBodies {
+  def apiMethod: Int = {
+    // as body of the method isn't part of the public interface, one can use all features of Scala here.
+    lazy val result = 0 // while lazy vals are prohibited in the class, they are allowed in the bodies of methods
+    result
+  }
+}
+{% endhighlight %}
+```
+
+## Features that will fail compilation with `@binaryCompatible`
+The features listed below have complex encodings that may change in future versions. We prefer not to compromise on them.
+Most of those features can be simulated in a binary compatible way by writing a verbose re-impelemtation 
+which won't rely on desugaring performed inside compiler.
+Note that while those features are prohibited in the public API, they can be safely used inside bodies of the methods.
+
+  - public fields. Can be simulated by explicitly defining public getters and setters that access a private field;
+  - lazy vals. Can be simulated by explicitly writing an implementation in source;
+  - case classes. Can be simulated by explicitly defining getters and other members synthesized for a case class(`copy`, `productArity`, `apply`, `unApply`, `unapply`).
+
+The features listed below cannot be easily re-implemented in a class or trait annotated with `@binaryCompatible`.
+  - default arguments;
+  - default methods. See Addendum;
+  - constant types(both explicit and inferred);
+  - inline.
```

Comment:
The idea is that `@stableABI` classes don't need `@scalaSignature` or `@tasty` to compile against. In order for inline to work, we'll need to be able to read the trees from those sections and make assumptions about those trees, that can be dependent on compiler version. 

I propose to go conservative route here, we'll have an opportunity to allow more features later, but we won't have an opportunity to remove them.